lib/fuse_lowlevel.c

```diff
@@ -1031,10 +1031,24 @@ int fuse_reply_poll(fuse_req_t req, unsigned revents)
 	return send_reply_ok(req, &arg, sizeof(arg));
 }
 
+/* Is `path` a single path component that is not "." or ".."? */
+static int is_safe_path_component(const char *path)
+{
+	if (strchr(path, '/'))
+		return 0;
+
+	return !is_dot_or_dotdot(path);
+}
+
 static void do_lookup(fuse_req_t req, fuse_ino_t nodeid, const void *inarg)
 {
 	char *name = (char *) inarg;
 
+	if (!is_safe_path_component(name)) {
+		fuse_reply_err(req, EINVAL);
```

Comment:
Actually lookup may receive "." or ".." from kernel when NFS export is enabled.  NFS exporting of fuse is not very good, and doesn't support generic file handles.   So longer term this interface probably needs to be replaced, but for now returning an error would be a regression.  This validation needs to be done in filesystems.

This only applies to LOOKUP, other ops should never get "." or ".." in file name.